### PR TITLE
For E0277 on `for` loops, point at the "head" expression

### PR DESCRIPTION
**Pull request by estebank**
When E0277's span points at a `for` loop, the actual issue is in the
element being iterated. Instead of pointing at the entire loop, point
only at the first line (when possible) so that the span ends in the
element for which E0277 was triggered.

### CODE REVIEW

File: src/librustc/traits/error_reporting.rs [hunk 1]
```diff
@@ -551,6 +551,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         let OnUnimplementedNote { message, label }
                             = self.on_unimplemented_note(trait_ref, obligation);
                         let have_alt_message = message.is_some() || label.is_some();
+                        let span = match self.tcx.sess.codemap().span_to_snippet(span) {
+                            Ok(ref _s) if _s.starts_with("for ") => {
```
Comment (zackmdavis) [inline src/librustc/traits/error_reporting.rs]: style nit: `Ok(ref s) if s.starts_with("for ")` ([pretty sure](https://play.rust-lang.org/?gist=3fd8dec472b33b02f0b34e92e1bdec15&version=stable) the unused-variables lint is in fact smart enough to count the `if` guard as a use)

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: For some reason I was convinced that there was a lint that complained at patterns that where only used in match guards... Must have been a different case.

Comment (zackmdavis) [reply src/librustc/traits/error_reporting.rs]: You might be thinking of how the dead-code lint complains about enum variants that are never constructed, even if they're matched on?

File: src/librustc/traits/error_reporting.rs [hunk 2]
```diff
@@ -551,6 +551,24 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                         let OnUnimplementedNote { message, label }
                             = self.on_unimplemented_note(trait_ref, obligation);
                         let have_alt_message = message.is_some() || label.is_some();
+                        let span = match self.tcx.sess.codemap().span_to_snippet(span) {
+                            Ok(ref s) if s.starts_with("for ") => {
+                                // On for loops, this error is caused by the element being iterated
```
Comment (nikomatsakis) [inline src/librustc/traits/error_reporting.rs]: Hmm, I like this span a lot, but I wonder if this is the right way to do this. Maybe we should adjust the span in HIR lowering instead? Roughly here:

https://github.com/rust-lang/rust/blob/a0dcecff90c45ad5d4eb60859e22bb3f1b03842a/src/librustc/hir/lowering.rs#L3167-L3173

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
I like the span but feel like we could do it in a more elegant way...

Also, what if people do something like

```
for x in
    something
{
    ...
}
```

I do that from time to time...

File: src/librustc/traits/error_reporting.rs (hunk 2 above)
Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: Reasonable, it's definitely an overuse of `span_until_char`, it was just slightly faster to figure how to do this than the proper way. Should it point at only the iterator expression or at the entire `for _ in iterator_expr`?

Comment (nikomatsakis) [reply src/librustc/traits/error_reporting.rs]: @estebank good question, I don't know the answer. Just pointing at the span of the iterator expression is obviously *easier*... and maybe better? I guess that showing the `for` may help convey that it's "because of the `for`" that `IntoIterator` is needed... but that seems sort of obvious...

If we do it in lowering, we can also introduce a "synthetic" span, right, that would let us give a very tailored message if we wanted?

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: > If we do it in lowering, we can also introduce a "synthetic" span, right, that would let us give a very tailored message if we wanted?

I think so. If we want to point at the `for`, we could just do that, which would look better ni the case where `for` and `expr` are not in the same line:

```
X |    for _ in expr {
  |    ---      ^^^^ `&str` is not an iterator
  |    |
  |    needed because of this `for` loop
```

Comment (nikomatsakis) [reply src/librustc/traits/error_reporting.rs]: oooh, that's nice

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: I looked at this and I have to rework the code to not lose the for-loop's context by the time the trait is being resolved. It might take me a while.

Comment (estebank) [reply src/librustc/traits/error_reporting.rs]: Modified the expansion code to use the appropriate span, but that makes it harder to add the appropriate span under `for`.

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -2969,7 +2969,7 @@ impl<'a> LoweringContext<'a> {
                     let body = if let Some(else_expr) = wildcard_arm {
                         P(self.lower_expr(else_expr))
                     } else {
-                        self.expr_tuple(e.span, hir_vec![])
+                        self.expr_tuple(body.span, hir_vec![])
```
Comment (nikomatsakis) [inline src/librustc/hir/lowering.rs]: Why the change to body.span here? It's not entirely intuitive to me that the (synthetic) "else clause" should get the span of the code that executes when `else` does not apply. =) At minimum, a comment seems good.

File: src/test/ui/const-fn-error.stderr [hunk 4]
```diff
@@ -13,22 +13,20 @@ error[E0016]: blocks in constant functions are limited to items and tail express
...
    |
-17 | /     for i in 0..x { //~ ERROR calls in constant functions
-18 | |     //~| ERROR constant function contains unimplemented
-19 | |         sum += i;
-20 | |     }
-   | |_____^
+17 |     for i in 0..x { //~ ERROR calls in constant functions
+   |              ^^^^
+   |              |
+   |              in this macro invocation
```
Comment (nikomatsakis) [inline src/test/ui/const-fn-error.stderr]: macro invocation? what macro? =)

Is this referring to the for loop desugaring?

File: src/test/ui/suggestions/for-c-in-str.rs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// E0277 should point exclusively at line 14, not the entire for loop span
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/for-c-in-str.rs]: nice

File: src/test/ui/const-fn-error.stderr (hunk 4 above)
Comment (estebank) [reply src/test/ui/const-fn-error.stderr]: I believe it is the desugaring of the range. Is it ok if I tackle this on a follow up PR? The span is probably pointing at the `x`.

File: src/librustc/hir/lowering.rs (hunk 3 above)
Comment (estebank) [reply src/librustc/hir/lowering.rs]: Reverted.

File: src/test/ui/const-fn-error.stderr (hunk 4 above)
Comment (estebank) [reply src/test/ui/const-fn-error.stderr]: Fixed.

File: src/test/ui/suggestions/try-operator-on-main.stderr [hunk 6]
```diff
@@ -2,10 +2,7 @@ error[E0277]: the `?` operator can only be used in a function that returns `Resu
   --> $DIR/try-operator-on-main.rs:19:5
    |
 19 |     std::fs::File::open("foo")?; //~ ERROR the `?` operator can only
-   |     ---------------------------
-   |     |
-   |     cannot use the `?` operator in a function that returns `()`
-   |     in this macro invocation
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/try-operator-on-main.stderr]: nice drive-by fix here 🚗

File: src/test/ui/suggestions/for-c-in-str.stderr [hunk 7]
```diff
@@ -0,0 +1,11 @@
+error[E0277]: the trait bound `&str: std::iter::Iterator` is not satisfied
+  --> $DIR/for-c-in-str.rs:14:14
+   |
+14 |     for c in "asdf" {
+   |              ^^^^^^ `&str` is not an iterator; maybe try calling `.iter()` or a similar method
+   |
+   = help: the trait `std::iter::Iterator` is not implemented for `&str`
+   = note: required by `std::iter::IntoIterator::into_iter`
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/for-c-in-str.stderr]: it'd be cool if we could highlight the `for`, but this is indeed pretty good still =)

**Review by nikomatsakis — APPROVED** [review]
this turned out really good

File: src/test/ui/suggestions/try-operator-on-main.stderr (hunk 6 above)
Comment (estebank) [reply src/test/ui/suggestions/try-operator-on-main.stderr]: 😎

File: src/test/ui/suggestions/for-c-in-str.stderr (hunk 7 above)
Comment (estebank) [reply src/test/ui/suggestions/for-c-in-str.stderr]: I agree. That will require some extra changes to `report_selection_error` in order to figure out that this is part of a `for`-loop desugaring in order to find the leading `for`. For now I think this is enough of an improvement.
